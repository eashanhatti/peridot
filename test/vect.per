define Nat
  : Type
  = Record {
      tag: Bool,
      data:
        if tag {
          Record {}
        } else {
          Nat
        }
    }

define zero
  : Nat
  = record {
      tag = true,
      data = record {}
    }

define succ
  : Function(Nat) -> Nat
  = function(n) =>
      record {
        tag = false,
        data = n
      }

define Vect
  : Function(Nat, Type) -> Type
  = function(n, A) =>
      Record {
        tag: Bool,
        data:
          if tag {
            Record {
              index: Equal(n, zero)
            }
          } else {
            Record {
              m: Nat,
              item: A,
              index: Equal(n, succ(m)),
              next: Vect(m, A)
            }
          }
      }

define nil
  : Function(inferred A: Type) -> Vect(zero, A)
  = function() =>
      record {
        tag = true,
        data = record { index = reflexive }
      }

define cons
  : Function(
      inferred A: Type,
      inferred n: Nat,
      item: A,
      vect: Vect(n, A))
      -> Vect(succ(n), A)
  = function(inferred A, inferred n, item, vect) =>
      record {
        tag = false,
        data =
          record {
            m = n,
            item = item,
            index = reflexive,
            next = vect
          }
      }
