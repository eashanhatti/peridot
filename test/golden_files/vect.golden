((fromList
    [(UVGlobal 1000,
      Just (Neutral _ (GlobalVar (Id {unId = 0})))),
     (UVGlobal 1001,
      Just (Neutral _ (GlobalVar (Id {unId = 0})))),
     (UVGlobal 1002,
      Just (Neutral _ (GlobalVar (Id {unId = 0})))),
     (UVGlobal 1003,
      Just (Neutral _ (GlobalVar (Id {unId = 0})))),
     (UVGlobal 1004,
      Just (Neutral _ (GlobalVar (Id {unId = 1})))),
     (UVGlobal 1005,
      Just
        (Neutral
           _
           (ObjFunElim
              (Neutral _ (GlobalVar (Id {unId = 2})))
              (LocalVar Level {unLevel = 1}))))],
  fromList [],
  fromList []),
 Let
   (fromList
      [ObjTerm
         (Id {unId = 0})
         (Rigid (TypeType Obj))
         (RecType
            (fromList
               [(Field {unField = "tag"}, Rigid TwoType),
                (Field {unField = "data"},
                 TwoElim
                   (LocalVar (Index {unIndex = 0}))
                   (RecType (fromList []))
                   (GlobalVar (Id {unId = 0})))])),
       ObjTerm
         (Id {unId = 1})
         (GlobalVar (Id {unId = 0}))
         (RecIntro
            (fromList
               [(Field {unField = "tag"}, Rigid TwoIntro0),
                (Field {unField = "data"},
                 RecIntro (fromList []))])),
       ObjTerm
         (Id {unId = 2})
         (ObjFunType
            Explicit
            (GlobalVar (Id {unId = 0}))
            (GlobalVar (Id {unId = 0})))
         (ObjFunIntro
            (RecIntro
               (fromList
                  [(Field {unField = "tag"}, Rigid TwoIntro1),
                   (Field {unField = "data"},
                    LocalVar (Index {unIndex = 1}))]))),
       ObjTerm
         (Id {unId = 3})
         (ObjFunType
            Explicit
            (GlobalVar (Id {unId = 0}))
            (ObjFunType
               Explicit
               (Rigid (TypeType Obj))
               (Rigid (TypeType Obj))))
         (ObjFunIntro
            (ObjFunIntro
               (RecType
                  (fromList
                     [(Field {unField = "tag"}, Rigid TwoType),
                      (Field {unField = "data"},
                       TwoElim
                         (LocalVar (Index {unIndex = 0}))
                         (RecType
                            (fromList
                               [(Field {unField = "index"},
                                 Rigid
                                   (ObjIdType
                                      (LocalVar (Index {unIndex = 2}))
                                      (GlobalVar (Id {unId = 1}))))]))
                         (RecType
                            (fromList
                               [(Field {unField = "m"},
                                 GlobalVar (Id {unId = 0})),
                                (Field {unField = "item"},
                                 LocalVar (Index {unIndex = 2})),
                                (Field {unField = "index"},
                                 Rigid
                                   (ObjIdType
                                      (LocalVar (Index {unIndex = 4}))
                                      (ObjFunElim
                                         (GlobalVar (Id {unId = 2}))
                                         (LocalVar (Index {unIndex = 1}))))),
                                (Field {unField = "next"},
                                 ObjFunElim
                                   (ObjFunElim
                                      (GlobalVar (Id {unId = 3}))
                                      (LocalVar (Index {unIndex = 2})))
                                   (LocalVar (Index {unIndex = 4})))])))])))),
       ObjTerm
         (Id {unId = 4})
         (ObjFunType
            Unification
            (Rigid (TypeType Obj))
            (ObjFunElim
               (ObjFunElim
                  (GlobalVar (Id {unId = 3}))
                  (GlobalVar (Id {unId = 1})))
               (LocalVar (Index {unIndex = 0}))))
         (ObjFunIntro
            (RecIntro
               (fromList
                  [(Field {unField = "tag"}, Rigid TwoIntro0),
                   (Field {unField = "data"},
                    RecIntro
                      (fromList
                         [(Field {unField = "index"},
                           Rigid (ObjIdIntro (UniVar (UVGlobal 1004))))]))]))),
       ObjTerm
         (Id {unId = 5})
         (ObjFunType
            Unification
            (Rigid (TypeType Obj))
            (ObjFunType
               Unification
               (GlobalVar (Id {unId = 0}))
               (ObjFunType
                  Explicit
                  (LocalVar (Index {unIndex = 1}))
                  (ObjFunType
                     Explicit
                     (ObjFunElim
                        (ObjFunElim
                           (GlobalVar (Id {unId = 3}))
                           (LocalVar (Index {unIndex = 1})))
                        (LocalVar (Index {unIndex = 2})))
                     (ObjFunElim
                        (ObjFunElim
                           (GlobalVar (Id {unId = 3}))
                           (ObjFunElim
                              (GlobalVar (Id {unId = 2}))
                              (LocalVar (Index {unIndex = 2}))))
                        (LocalVar (Index {unIndex = 3})))))))
         (ObjFunIntro
            (ObjFunIntro
               (ObjFunIntro
                  (ObjFunIntro
                     (RecIntro
                        (fromList
                           [(Field {unField = "tag"}, Rigid TwoIntro1),
                            (Field {unField = "data"},
                             RecIntro
                               (fromList
                                  [(Field {unField = "m"},
                                    LocalVar (Index {unIndex = 3})),
                                   (Field {unField = "item"},
                                    LocalVar (Index {unIndex = 3})),
                                   (Field {unField = "index"},
                                    Rigid
                                      (ObjIdIntro (UniVar (UVGlobal 1005)))),
                                   (Field {unField = "next"},
                                    LocalVar (Index {unIndex = 4}))]))]))))))])
   (Rigid (TypeType Obj)))
